src/transform/src/predicate_pushdown.rs

```diff
@@ -187,10 +187,7 @@ impl PredicatePushdown {
                         let mut pushed = false;
                         // Attempt to push down each predicate to each input.
                         for (index, push_down) in push_downs.iter_mut().enumerate() {
-                            if let RelationExpr::ArrangeBy { .. } = inputs[index] {
-                                // do nothing. We do not want to push down a filter and block
-                                // usage of an index
```

Comment:
this is ok to remove because join planning is capable of lifting filters out of the way in order to use an arrangement.

Also, removing this prevents filters not involved in the semijoin from becoming equivalence classes in the join.

Reply:
> removing this prevents filters not involved in the semijoin from becoming equivalence classes in the join.

I don't understand this!

Reply:
Take this scenario:

1) There is an index on `foo(a)`. 
2) the query is `SELECT b, c FROM foo WHERE a = 5 and b = 'this'`

The `FilterEqualLiteral` changes the filter on column `a` to a join, so the RelationExpr looks roughly like this:
```
%0 =
| Get materialize.public.foo (u1)
| ArrangeBy (#0)

%1 =
| Join %0 (= #0 5)
| Filter (#1 = "this")
| Project (#1, #2)
```

The logic `PredicatePushdown` for `Filter{Join{...}}` works like this:
1) Try to localize the predicate to a single input
2) If the predicate localization fails, then push the predicate into the equivalence class.

If I don't delete these lines, localizing `Filter(#1 = this)` automatically fails, so the predicate gets pushed into the equivalence class of the Join like this.
```
%0 =
| Get materialize.public.foo (u1)
| ArrangeBy (#0)

%1 =
| Join %0 (= #0 5) (#1 = "this")
| Project (#1, #2)
```

Reply:
I'm sorry, I still don't understand. The plan you have looks great to me. Is the problem that the semijoin implementation doesn't handle additional filters?

Reply:
A better example of what happens if the lines are not removed. Equality filters that don't have a literal on one side get pushed into the join, and then their presence messes up join implementation planning.

```
# indexes exist on foo(b) and foo(a, b, c)
query T multiline
EXPLAIN PLAN FOR SELECT a FROM foo WHERE b = 'it' and a = c
----
%0 =
| Get materialize.public.foo (u1)
| ArrangeBy (#0, #1, #2)

%1 =
| Join %0 (= #0 #2) (= #1 "it")
| | implementation = Differential %0.(#0, #1, #2)
| | demand = (#0)
| Project (#0)

EOF
```

```
# index exists on foo(b)
query T multiline
EXPLAIN PLAN FOR SELECT a, c FROM foo WHERE b = 'it' and a = c
----
%0 =
| Get materialize.public.foo (u1)

%1 =
| Join %0 (= #0 #2) (= #1 "it")
| | implementation = Differential %0
| | demand = (#0)
| Project (#0)

EOF
```